crypto/keys/ecdsa/pubkey.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/cosmos/cosmos-sdk/types/address"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// signature holds the r and s values of an ECDSA signature.
+type signature struct {
+	R, S *big.Int
+}
+
+type ecdsaPK struct {
+	ecdsa.PublicKey
+
+	// cache
+	address tmcrypto.Address
+}
+
+// String implements PubKey interface.
+func (pk *ecdsaPK) Address() tmcrypto.Address {
+	if pk.address == nil {
+		pk.address = address.Hash(curveNames[pk.Curve], pk.Bytes())
+	}
+	return pk.address
+}
+
+// Bytes returns the byte representation of the public key using a compressed form
+// specified in section 4.3.6 of ANSI X9.62 with first byte being the curve type.
+func (pk *ecdsaPK) Bytes() []byte {
+	if pk == nil {
+		return nil
+	}
+	compressed := make([]byte, PubKeySize)
+	compressed[0] = curveTypes[pk.Curve]
+	compressed[1] = byte(pk.Y.Bit(0)) | 2
+	pk.X.FillBytes(compressed[2:])
+	return compressed
+}
+
+// Equals - you probably don't need to use this.
+// Runs in constant time based on length of the keys.
+func (pk *ecdsaPK) Equals(other cryptotypes.PubKey) bool {
+	pk2, ok := other.(*ecdsaPK)
+	if !ok {
+		return false
+	}
+
+	return pk.PublicKey.Equal(&pk2.PublicKey)
+}
+
+// VerifySignature implements skd.PubKey interface.
+func (pk *ecdsaPK) VerifySignature(msg []byte, sig []byte) bool {
+	s := new(signature)
+	if _, err := asn1.Unmarshal(sig, s); err != nil || s == nil {
+		return false
+	}
+
+	h := sha256.Sum256(msg)
+	return ecdsa.Verify(&pk.PublicKey, h[:], s.R, s.S)
+}
+
+// Type returns key type name. Implements sdk.PubKey interface.
+func (pk *ecdsaPK) Type() string {
+	return curveNames[pk.Curve]
+}
+
+// **** proto.Message ****
+
+// Reset implements proto.Message interface.
+func (pk *ecdsaPK) Reset() {
+	*pk = ecdsaPK{}
+}
+
+// ProtoMessage implements proto.Message interface.
+func (*ecdsaPK) ProtoMessage() {}
+
+// String implements proto.Message interface
+func (pk *ecdsaPK) String() string {
+	return fmt.Sprintf("%s{%X}", curveNames[pk.Curve], pk.Bytes())
+}
+
+// **** Proto Marshaler ****
+
+// Marshal implements ProtoMarshaler interface.
+func (pk *ecdsaPK) Marshal() ([]byte, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.Marshal()
+}
+
+// MarshalTo implements ProtoMarshaler interface.
+func (pk *ecdsaPK) MarshalTo(data []byte) (int, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.MarshalTo(data)
+}
+
+// MarshalToSizedBuffer implements ProtoMarshaler interface.
+func (pk *ecdsaPK) MarshalToSizedBuffer(dAtA []byte) (int, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.MarshalToSizedBuffer(dAtA)
+}
+
+// Unmarshal implements ProtoMarshaler interface.
+func (pk *ecdsaPK) Unmarshal(b []byte) error {
+	bv := gogotypes.BytesValue{}
+	err := bv.Unmarshal(b)
+	if err != nil {
+		return err
+	}
+	return pk.UnmarshalAmino(bv.Value)
+}
+
+// Size implements ProtoMarshaler interface.
+func (pk *ecdsaPK) Size() int {
+	if pk == nil {
+		return 0
+	}
+	return sovPubKeySize
+}
```

Comment:
These methods should never be implemented manually. There needs to be a proto file and generated code.

Reply:
I'm using `gogotypes.BytesValue` to carry the proto format. 
I started with an approach of creating a proto files for pub and public keys - but that was too much overhead:
* we would need to crate proxy types for big.Int and do the serialization for it as well.
* more storage overhead - in the solution above I'm using a compact representation.
* using proxy and convertions between types would also add additional CPU cycles.